src/mux_client_impl/command.rs

```diff
@@ -0,0 +1,117 @@
+use super::Error;
+use super::RemoteChild;
+use super::{as_raw_fd, ChildStderr, ChildStdin, ChildStdout, Stdio};
+
+use std::ffi::OsStr;
+use std::path::PathBuf;
+use std::process;
+
+use openssh_mux_client::connection::{Connection, EstablishedSession, Session};
+
+#[derive(Debug)]
+pub struct Command {
+    cmd: String,
+    ctl: PathBuf,
+
+    stdin_v: Stdio,
+    stdout_v: Stdio,
+    stderr_v: Stdio,
+}
+
+impl Command {
+    pub(crate) fn new(ctl: PathBuf, cmd: String) -> Self {
+        Self {
+            cmd,
+            ctl,
+
+            stdin_v: Stdio::null(),
+            stdout_v: Stdio::null(),
+            stderr_v: Stdio::null(),
+        }
+    }
+}
+
+impl Command {
+    pub fn arg<S: AsRef<str>>(&mut self, arg: S) -> &mut Self {
+        self.cmd.push_str(" \"");
+        self.cmd.push_str(arg.as_ref());
+        self.cmd.push('"');
+        self
+    }
+
+    pub fn raw_arg<S: AsRef<OsStr>>(&mut self, arg: S) -> &mut Self {
+        self.cmd.push(' ');
+        self.cmd.push_str(&arg.as_ref().to_string_lossy());
+        self
+    }
+
+    pub fn stdin<T: Into<Stdio>>(&mut self, cfg: T) -> &mut Self {
+        self.stdin_v = cfg.into();
+        self
+    }
+
+    pub fn stdout<T: Into<Stdio>>(&mut self, cfg: T) -> &mut Self {
+        self.stdout_v = cfg.into();
+        self
+    }
+
+    pub fn stderr<T: Into<Stdio>>(&mut self, cfg: T) -> &mut Self {
+        self.stderr_v = cfg.into();
+        self
+    }
+
+    /// After this function, stdin, stdout and stderr is reset.
+    async fn spawn_impl(
+        &mut self,
+    ) -> Result<
+        (
+            EstablishedSession,
+            Option<ChildStdin>,
+            Option<ChildStdout>,
+            Option<ChildStderr>,
+        ),
+        Error,
+    > {
+        let (stdin, child_stdin) = self.stdin_v.take().into_stdin()?;
+        let (stdout, child_stdout) = self.stdout_v.take().into_stdout()?;
+        let (stderr, child_stderr) = self.stderr_v.take().into_stderr()?;
+
+        // Then launch!
+        let session = Session::builder().cmd(&self.cmd).build();
+
+        let established_session = Connection::connect(&self.ctl)
+            .await?
+            .open_new_session(
+                &session,
+                &[as_raw_fd(&stdin), as_raw_fd(&stdout), as_raw_fd(&stderr)],
+            )
+            .await?;
+
+        Ok((established_session, child_stdin, child_stdout, child_stderr))
+    }
+
+    pub async fn spawn(&mut self) -> Result<RemoteChild, Error> {
+        let (established_session, child_stdin, child_stdout, child_stderr) =
+            self.spawn_impl().await?;
+
+        Ok(RemoteChild::new(
+            established_session,
+            child_stdin,
+            child_stdout,
+            child_stderr,
+        ))
+    }
+
+    pub async fn output(&mut self) -> Result<process::Output, Error> {
+        self.stdout(Stdio::piped())
```

Comment:
As mentioned in an earlier comment, this doesn't seem quite right, as it'll override any choices the user made for these streams.

Reply:
This function here is actually unused, so I will clean them and remove these unused functions.

Reply:
I have removed all unused functions in `process_impl` and `mux_client_impl`, now they should be easier to review.